### PR TITLE
[SPARK-27398][SQL] Refactoring of CreateJacksonParser.getStreamDecoder

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to simplify the `CreateJacksonParser.getStreamDecoder` method, and use `InputStreamReader` instead of creating `ReadableByteChannel` and do the channel decoding via `StreamDecoder.forDecoder`. Also I replaced returned type by `Reader` and renamed the method from `getStreamDecoder` to `getReader`.

## How was this patch tested?

by `JsonSuite`, `JsonFunctionsSuite` and `JsonExpressionsSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
ping @hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala [hunk 1]
```diff
@@ -55,19 +52,13 @@ private[sql] object CreateJacksonParser extends Serializable {
...
-  // decoding buffer according to size of input array.
-  private def getStreamDecoder(enc: String, in: Array[Byte], length: Int): StreamDecoder = {
+  private def getReader(enc: String, in: Array[Byte], length: Int): Reader = {
     val bais = new ByteArrayInputStream(in, 0, length)
-    val byteChannel = Channels.newChannel(bais)
-    val decodingBufferSize = Math.min(length, 8192)
-    val decoder = Charset.forName(enc).newDecoder()
-
-    StreamDecoder.forDecoder(byteChannel, decoder, decodingBufferSize)
+    new InputStreamReader(bais, enc)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala]: Seems fine but I'm wondering why it used channels in the first place. Was it for performance?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/CreateJacksonParser.scala]: I took the code from internals of Jackson parser, and kept it as is to be sure that the behavior is the same. `InputStreamReader` uses similar code but without creating a channel. So, I think we could use the standard API of `InputStreamReader`.

**Comment by MaxGekk** [comment]
I re-ran JSON benchmark, and unfortunately it shows performance regression up to 2 times. For example, the last benchmarks:
Before:
```
son files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Text read                                          7537           7556          26          6.6         150.7       1.0X
Schema inferring                                  27875          28306         499          1.8         557.5       0.3X
Parsing without charset                           26030          26083          67          1.9         520.6       0.3X
Parsing with UTF-8                                37115          37480         392          1.3         742.3       0.2X
```
After:
```
Json files in the per-line mode:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Text read                                          7435           7457          27          6.7         148.7       1.0X
Schema inferring                                  28307          28378          70          1.8         566.1       0.3X
Parsing without charset                           25104          25197          89          2.0         502.1       0.3X
Parsing with UTF-8                                66200          66402         216          0.8        1324.0       0.1X
```
I am closing the PR.

**Pull request closed without merging**